File: src/Domain/Entities/Restrictions/Restriction.cs
using Crpg.Domain.Common;
using Crpg.Domain.Entities.Users;

namespace Crpg.Domain.Entities.Restrictions;

/// <summary>
/// Represents a user restriction. Used by a game server to know if a user is allowed to play/chat.
/// </summary>
public class Restriction : AuditableEntity
{
    public int Id { get; set; }
    public int RestrictedUserId { get; set; }

    /// <summary>
    /// Duration of the restriction. <see cref="TimeSpan.Zero"/> to cancel the past restriction.
    /// </summary>
    /// <remarks>Add it <see cref="AuditableEntity.CreatedAt"/> to get the end of the restriction.</remarks>
    public TimeSpan Duration { get; set; }
    public RestrictionType Type { get; set; }
    public string Reason { get; set; } = string.Empty;
    public int RestrictedByUserId { get; set; }

    public User? RestrictedUser { get; set; }
    public User? RestrictedByUser { get; set; }
}

File: src/WebUI/src/models/restriction.ts
import { UserPublic } from '@/models/user';

export enum RestrictionType {
  Join = 'Join',
  Chat = 'Chat',
  ALL = 'All',
}

export interface Restriction {
  id: number;
  restrictedUser: UserPublic;
  duration: number;
  type: RestrictionType;
  reason: string;
  restrictedByUser: UserPublic;
  createdAt: Date;
}

export interface RestrictionWithActive extends Restriction {
  active: boolean;
}

export interface RestrictionCreation {
  restrictedUserId: number;
  type: RestrictionType;
  reason: string;
  duration: number;
}
